base/c.jl

```diff
@@ -503,3 +503,183 @@ end
+"""
+function getargs(call)
+    # no arguments
+    length(call.args) < 2 &&
+        return [], []
+
+    firstarg = call.args[2]
+    hashead(firstarg, :parameters) &&
+        return call.args[3:end], firstarg.args
+
+    return call.args[2:end], []
+end
+
+"""
+       mkarg(arg)
+
+`mkarg` takes an argument and returns a tuple of symbol and type. The
+argument must have a type annotation or an error will be thrown.
+"""
+function mkarg(arg)
+    !hashead(arg, :(::)) &&
+        throw(CcallError("args in @ccall need type annotations. '$arg' doesn't have one."))
+    return (arg=arg.args[1], type=arg.args[2])
+end
+
+"""
+get the type specified on the varargs. throw if there is more than one type.
+"""
+function getvarargtype(varargs)
+    vararg_types = Set(a.type for a in varargs)
+    length(vararg_types) > 1 &&
+        throw(CcallError("varargs @ccall with different argument types not yet supported"))
+    return pop!(vararg_types)
+end
+
+"""
+    parsecall(expression)
+
+`parsecall` is an implementation detail of `@ccall
+
+it takes and expression like `:(printf("%d"::Cstring, value::Cuint)::Cvoid)`
+returns: a tuple of `(function_name, return_type, arg_types, args)`
+
+The above input outputs this:
+
+    (:printf, :Cvoid, :((Cstring, Cuint)), ["%d", :value])
+
+Note that the args are in an array, not a quote block and have to be
+appended to the ccall in a separate step.
+"""
+function parsecall(expr::Expr)
+    # setup and check for errors
+    !hashead(expr, :(::)) && begin
+        throw(CcallError("@ccall needs a function signature with a return type"))
+    end
+    rettype = expr.args[2]
+
+    call = expr.args[1]
+    !hashead(call, :call) &&
+        throw(CcallError("@ccall has to take a function call"))
+
+    # get the function symbols
+    func = getcfunc(call.args[1])
+    normalargs, varargs = getargs(call)
+
+
+    # separate annotations from names
+    normalargs = mkarg.(normalargs)
+    args = Any[a.arg for a in normalargs]
+    types = Any[a.type for a in normalargs]
+    argtypes = :(())
+    argtypes.args = types
+
+    isempty(varargs) && return func, rettype, argtypes, args, 0
+    nreq = length(args)
+
+    # vararg handling.
+    varargs = mkarg.(varargs)
+    append!(args, (a.arg for a in varargs))
+    append!(argtypes.args, (a.type for a in varargs))
+    return func, rettype, argtypes, args, nreq
+end
+
+function lower(convention, func, rettype, argtypes, args, nreq)
+    lowering = []
+    realargs = []
+    gcroots = []
+    for (i, (arg, type)) in enumerate(zip(args, argtypes.args))
+        sym = Symbol(string("%", i))
+        sym2 = Symbol(string("%", i + length(args)))
+        earg, etype = esc.([arg, type])
+        push!(lowering, :($sym = Base.cconvert($etype, $earg)))
+        push!(lowering, :($sym2 = Base.unsafe_convert($etype, $sym)))
+        push!(realargs, sym2)
+        push!(gcroots, sym)
+    end
+    etypes = :(Core.svec())
+    append!(etypes.args, argtypes.args)
```

Comment:
`Core.svec` (`Base.getproperty(Core, :svec)`) is very different. You want `Expr(:call, Expr(:core, :svec), argtypes.args...)` (equivalently, `GlobalRef(Core, :svec)`)

Reply:
I did wonder about this and was about to suggest lowering to `GlobalRef(Core, :svec)`.

But isn't the `Core.svec` binding evaluated in the module of the macro definition rather than the user code, provided the `esc`'s are in the right place? In that case it should be fine to use the symbolic form `Core.svec` (`Expr(:., :Core, QuoteNode(:svec)))`)?

Reply:
fixed in https://github.com/JuliaLang/julia/pull/32748/commits/0ae5b41b691db050b753b8f80c1fe38541805f35 (using `Expr(:call, Expr(:core, :svec),`...)